### PR TITLE
unbound: add odhcpd scripts and act like dnsmasq without dnsmasq

### PR DESCRIPTION
**Pull request by EricLuehrsen**
Maintainer: me
Compile tested: LEDE trunk, OpenWrt CC
Run tested: TP Link WDR3600 and Archer C7
Ran these two together to test how they resolved with respect to each other, separate static routed subnets, and various settings. In particular, it is important to see that the router name was properly resolved. LuCI to any router for maintenance from anywhere by name. Seems fairly solid.

Description:
Unbound+DHCP (server of your choice) should be able to replicate
a lot of what dnsmasq provides. With this change set Unbound
still works with dnsmasq, but also it can work with a plain
DHCP server. Features have been added within the UCI itself
to act like dnsmasq.

- alone: name each interface relative to router hostname
- alone: prevent upstream leakage of your domain and '.local'
- dnsmasq: use dnsmasq UCI to configure forwarding clauses
- dhcp: work with odhcpd as example of companion DHCP-DNS
- dhcp: convert DHCPv4 leases into EUI64 SLAAC for DNS records
- all: enable encrypted remote unbound-control using splice conf
- all: allow user spliced conf-files for hybrid UCI and manual conf
-- 'unbound_srv.conf' will be spliced into the 'server:' clause
-- 'unbound_ext.conf' will add clauses to the end, example 'forward:'

README HOW TO for dnsmasq-in-serial, dnsmasq-in-parallel, and
unbound-with-odhcpd have better/added UCI starters. HOW TO for
including unbound_srv.conf and unbound_ext.conf are added.
Document new UCI: add_local_fqdn, add_wan_fqdn, dhcp4_slaac6,
dhcp_link, domain, and domain_type

@Borromini @hansmi @danrl @dibdot 

Luci:
https://github.com/openwrt/luci/pull/959

Reference (not dependent):
https://github.com/lede-project/source/pull/704

### CODE REVIEW

File: net/unbound/files/README.md [hunk 1]
```diff
@@ -8,82 +8,170 @@ Unbound may be useful on consumer grade embedded hardware. It is *intended* to b
...
+
+	config odhcpd 'odhcpd'
+		option maindhcp '1'
+		option leasefile '/var/lib/odhcpd/dhcp.leases'
+		# this is where the magic happens
+		option leasetrigger '/usr/lib/unbound/odhcpd.sh'
 
 ## Back to Manual Configuration
-You don't want UCI, but don't worry. We have UCI for that. However, OpenWrt or LEDE are targeted at embedded machines with flash ROM. The initialization scripts do a few things to protect flash ROM. 
+You do not want UCI. We have UCI for that. However, OpenWrt or LEDE are targeted at embedded machines with flash ROM. The initialization scripts do a few things to protect flash ROM.
```
Comment (danrl) [inline net/unbound/files/README.md]: First two sentences are hard to grasp. I do not want UCI but we have it for it? Maybe reword?

File: net/unbound/files/README.md [hunk 2]
```diff
@@ -8,82 +8,170 @@ Unbound may be useful on consumer grade embedded hardware. It is *intended* to b
...
+		list server '::1#1053'
+		...
+
+### Parallel dnsmasq
+In this case, Unbound serves your local network directly for all purposes. It will look over to dnsmasq for DHCP-DNS resolution. Unbound is generally accessible on port 53, and dnsmasq is only accessed at 127.0.0.1:1053 by Unbound. Although you can dig/drill/nslookup remotely with the proper directives.
 
 **/etc/config/unbound**:
 
 	config unbound
-		option dnsmasq_link_dns '1'
```
Comment (Borromini) [inline net/unbound/files/README.md]: So from what I can gather dnsmasq_link_dns is going away altogether? Is there any setting replacing this or is it now default behaviour with unbound linking to dnsmasq for local DHCP-DNS?

Comment (EricLuehrsen) [inline net/unbound/files/README.md]: Please use `option dhcp_link 'dnsmasq'` which replaces the old boolean with a scalable multi program state. Also the UCI will interogate dnsmasq UCI for the other options. This so you dont have different options in different UCI that need to match.

File: net/unbound/files/README.md (hunk 1 above)
Comment (EricLuehrsen) [inline net/unbound/files/README.md]: Yes. I am actively chopping the documentation down. It may get uglier before it gets better.

Comment (EricLuehrsen) [inline net/unbound/files/README.md]: And LuCI needs a matched update. Almost done with that, but it needs some arrangement or style or something.

**Comment by poranje** [comment]
beautiful

File: net/unbound/files/README.md (hunk 2 above)
Comment (Borromini) [inline net/unbound/files/README.md]: OK. I built 1.6.0-3, installed it over my existing -2 install, the opkg configuration file still contains a dnsmasq_link_dns option though - an oversight?

I am not using odhcpd so won't be able to comment on that.

Comment (EricLuehrsen) [inline net/unbound/files/README.md]: Added uci primer yesterday but: (1) "squash" and `git rebase upstream/master` require a `git push --force` and the SDK will fail/skip package updates `./scripts/feeds update -a` unless you take other actions like `rm -fR feeds/packages`. (2) UCI files are listed in the special configuration install target, and opkg should prompt you to keep or replace.

File: net/unbound/files/odhcpd.sh [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      cat $dns_ls_del | $UNBOUND_CONTROL_CFG local_datas_remove
+    fi
+    
+
+    if [ -f "$dns_ls_add" ] ; then
+      cat $dns_ls_add | $UNBOUND_CONTROL_CFG local_datas
+    fi
+    
+    
+    # prepare next round
```
Comment (danrl) [inline net/unbound/files/odhcpd.sh]: Quick question: Does this mean this code runs in some sort of loop? Or is this action triggered by an event?

**Comment by danrl** [comment]
I built an image without dnsmasq and using unbound+odhcpd only. It works fine for me. I have not tested the dynamic host entries due to lack of time.

Looking forward to see the changes in master!

File: net/unbound/files/odhcpd.sh (hunk 3 above)
Comment (EricLuehrsen) [inline net/unbound/files/odhcpd.sh]: odhpcd calls `leasetrigger` on renew, rebind, confirm, and release (I think I have that right). So in a busy network this script runs as if its on a loop, but it is really event based.

Comment (EricLuehrsen) [inline net/unbound/files/odhcpd.sh]: Side note: unlike other dhcp servers odhcpd doest send any parameters to the script. thats why the code to diff the lease file.

Comment (danrl) [inline net/unbound/files/odhcpd.sh]: Thanks for the clarification. The code is really beautiful.

File: net/unbound/files/README.md (hunk 2 above)
Comment (Borromini) [inline net/unbound/files/README.md]: OK so I probably pulled before you added the uci changes?

Opkg informed me about the config files differing.

**Comment by EricLuehrsen** [comment]
@thess - Ready to merge when you have time. Thank You.

**Comment by EricLuehrsen** [comment]
@hnyman if your back patrolling the endless flow of pull requests and have the time, then this one is ready to merge. Thank you.

**Comment by cshoredaniel** [comment]
On Sun, 22 Jan 2017 17:50:56 -0800
Eric Luehrsen <notifications@github.com> wrote:

> @hnyman if your back patrolling the endless flow of pull requests and
> have the time, then this one is ready to merge. Thank you.
> 

@hnyman, also if you and other packages feed folks are willing to put
up with my efforts to help (and as with the tags thing, send a little
note when I need to make corrections I reasonably can; I'm not sure I
know how to be as terse as everyone else though).

Regards,

Daniel

**Comment by thess** [comment]
@cshoredaniel - I think you are doing a fine job with the LuCi repo. However I, for one, would appreciate it if you didn't go all nuts with tags here. I have no trouble following stuff with the current set and really do not need more. 

@hnyman - It would have been nice if we had this submission squashed and cleaned up a bit before merge. Sigh...

**Comment by hnyman** [comment]
Well, the commits seemed to be topic-based, so I merged them as they were.

And have you checked Luci repo today for tags? Still too nuts? I gave similar feedback to cshoredaniel in https://github.com/openwrt/luci/issues/968 and now the tag usage is much more conservative than two days ago.

**Comment by thess** [comment]
Ah yes, much better :smile:
It no longer looks like a box of crayons.

**Comment by EricLuehrsen** [comment]
Yeah its always a tough balance with big commits. The last two maybe could have been squashed but no one asked. However I do strongly support bug fixes, new files (yet unused), and in-use enhancements as separate commits. Its a matter of change ( and regression) management.

- Eric


-------- Original message --------

**Comment by cshoredaniel** [comment]
@thess @hnyman  I actually did for searching and filtering rather than presentation (I guess I've dealt with too many things where visual inspection wasn't practical).

**Comment by poranje** [comment]
With unbound uci option dhcp_link set to 'odhcpd':

Any clue on how to handle hostnames for hosts that are having static IP configurations (so not using a DHCP(6) client) ?
For this purpose dnsmasq processes the uci stances "dhcp.domain", but unbound seems to ignore those. Writing records in /etc/hosts may a way, but using uci would be preferred.

**Comment by poranje** [comment]
With unbound uci option dhcp_link set to 'dnsmasq':

lsof shows that dnsmasq listens on port 1053 at all interfaces (even at global scope IPv6) and not only at the loopback interface. Since dnsmasq resolves only queries forwarded by unbound, listening on lo would suffice, so why does it listen on all interfaces ?

**Comment by EricLuehrsen** [comment]
This is the default behavior of dnsmasq on LEDE. It binds to `--port` on the wildcard interface. The Unbound package UCI does not and will not interfere with another package UCI. That is Read only. dnsmasq UCI has the ability to control this, and that is the choice of local administration. For example it may be useful within the lan to `drill -p 1053 ....` to test just dnsmasq remotely.

**Comment by EricLuehrsen** [comment]
> With unbound uci option dhcp_link set to 'dnsmasq':

This is the default behavior of dnsmasq on LEDE. It binds to `--port` on the wildcard interface.  dnsmasq UCI has the ability to control this, and that is the choice of local administration. For example it may be useful within the lan to `drill -p 1053 ....` to test just dnsmasq remotely. I have tried to be careful, so Unbound package UCI does not and will not interfere with another package UCI. That is external UCI is read only.

**Comment by EricLuehrsen** [comment]
> With unbound uci option dhcp_link set to 'odhcpd':

please pull a new github issue. Expand on your desired function and include an example use case. I will work out a robust and flexible method. It might take time to mimic dnsmasq UCI as close as possible. Its a continued goal to make an Unbound/dnsmasq swap a low learning curve.

**Comment by poranje** [comment]
created issue #4328
Paul
